planner/cardinality/histogram.go

```diff
@@ -0,0 +1,194 @@
+package cardinality
+
+import (
+	"bytes"
+
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/mysql"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/statistics"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/ranger"
+	"go.uber.org/zap"
+)
+
+// NewHistCollBySelectivity creates new HistColl by the given statsNodes.
+// TODO: remove this function after removing docs.pingcap.com/tidb/stable/system-variables#tidb_optimizer_selectivity_level.
+func NewHistCollBySelectivity(sctx sessionctx.Context, coll *statistics.HistColl, statsNodes []*StatsNode) *statistics.HistColl {
+	newColl := &statistics.HistColl{
+		Columns:       make(map[int64]*statistics.Column),
+		Indices:       make(map[int64]*statistics.Index),
+		Idx2ColumnIDs: coll.Idx2ColumnIDs,
+		ColID2IdxIDs:  coll.ColID2IdxIDs,
+		RealtimeCount: coll.RealtimeCount,
+	}
+	for _, node := range statsNodes {
+		if node.Tp == IndexType {
+			idxHist, ok := coll.Indices[node.ID]
+			if !ok {
+				continue
+			}
+			newIdxHist, err := newIndexBySelectivity(sctx.GetSessionVars().StmtCtx, idxHist, node)
+			if err != nil {
+				logutil.BgLogger().Warn("something wrong happened when calculating row count, "+
+					"failed to build histogram for index %v of table %v",
+					zap.String("category", "Histogram-in-plan"), zap.String("index", idxHist.Info.Name.O),
+					zap.String("table", idxHist.Info.Table.O), zap.Error(err))
+				continue
+			}
+			newColl.Indices[node.ID] = newIdxHist
+			continue
+		}
+		oldCol, ok := coll.Columns[node.ID]
+		if !ok {
+			continue
+		}
+		newCol := &statistics.Column{
```

Comment:
We can add NewColumn for it.

Reply:
That's exactly what I want to do later on... Use a standard NewFunc to create these structures.